### PR TITLE
✨ Prev & next post filtering, with primary tag support

### LINKED ISSUE #9140: Prev & Next 2.0: refactor & primary tag support
## Part 1: Refactor next/prev

Many, many moons ago, we added special helpers `{{prev_post}}` and `{{next_post}}` to the theme API.

You can read a little bit about how we struggled to implement this, back in the day before we had a proper API, or a plan, or well anything [here](https://github.com/TryGhost/Ghost/issues/4262) and [here](https://github.com/TryGhost/Ghost/issues/4799).

At the time, we ending up implementing this as a set of ?include parameters that could be fetched from the posts.read endpoint, i.e. you could fetch one post WITH its next and/or previous post at the same time. This was inefficient, and results in lots of extra queries.

Now in Ghost, we have a feature-rich `filter` parameter on the posts API, which makes implementing  something like next and previous posts much easier. Even though our documentation for the [get helper](https://themes.ghost.org/docs/get) officially lists date filtering as being one of the remaining limitations (because I'd never tested it) it does actually work 🎉  apparently, across sqlite3 and mysql without issue, and really quite well.

Previous post query: `/posts/?filter=slug:-[current-post-slug]+published_at:<='current-post-published-at'&order=published_at desc&limit=1`
Next post query: `/posts/?filter="slug:-[current-post-slug]+published_at:>='current-post-published-at'&order=published_at asc&limit=1`

Seeing as the old implementation is inefficient, ugly and results in a lot of extra code, and implementing this with filters is efficient and very light on code, we want to get rid of the old implementation.

This would be a breaking API change, however the ability to include next/prev was never documented, and I do not see any usages of it on Ghost(Pro) in the last 7 days, and therefore I believe it is only ever used via the `{{next_post}}` and `{{prev_post}}` helpers, and is safe to remove.

## Part 2: Add support for primary tags

Using the new, filter based implementation, we'd like to make it possible to output the next or previous post, which have the same primary tag as the current post.

This will be implemented with an extra parameter for the `{{next_post}}` and `{{prev_post}}` and helpers, e.g:

```
{{#next_post in="primary_tag"}}
      {{> "post-card"}}
{{/next_post}}

{{#prev_post in="primary_tag"}}
       {{> "post-card"}}
{{/prev_post}}
```

This will change which next and previous post is output, so that it is the next post inside the list of posts which share the same primary tag, instead of the next post across the whole blog.

If you don't provide the `in=""` clause, the `{{#next_post}}{{/next_post}}` & `{{#prev_post}}{{/prev_post}}`  helpers will continue to work exactly as they do now.

Short term, the only value the `in=""` clause will support is "primary_tag". Longer term, we'll roll out support for channels when #5091 is closer to being usable, etc.

-----

# Summary

This issue proposes a breaking change to the JSON API, removing the undocumented ability to fetch next/prev post directly using includes. Instead, anyone who did use this feature would need to use the filter param instead.

Meanwhile, the widely used `{{#next_post}}{{/next_post}}` & `{{#prev_post}}{{/prev_post}}` helpers will work exactly as they did before, with no changes. The underlying implementation will be more efficient (hopefully improving page load times and reducing load on the database).

Finally, we're adding an extra parameter to the `{{#next_post}}{{/next_post}}` & `{{#prev_post}}{{/prev_post}}` to support fetching the next/prev post inside of a primary tag.

- [x] Remove old next/prev includes from post API
- [x] Swap the `{{#next_post}}{{/next_post}}` & `{{#prev_post}}{{/prev_post}}` helpers to use the filter parameter instead
- [x] Tests for the API & date based filtering
- [ ] Update get helper documentation to include date-based filtering
- [ ] Update api documentation to  include data-based filtering

### PR DESCRIPTION
**Pull request by ErisDS**
Replace prev & next logic with the filter param, add new "in" option to support next and preivous post within primary tags.

closes #9140

---

This PR introduces a breaking change to the JSON API, removing the undocumented ability to fetch next/prev post directly using includes. Instead, anyone who did use this feature would need to use the filter param instead:

Previous post query: `/posts/?filter=slug:-[current-post-slug]+published_at:<='current-post-published-at'&order=published_at desc&limit=1`
Next post query: `/posts/?filter="slug:-[current-post-slug]+published_at:>='current-post-published-at'&order=published_at asc&limit=1`

Meanwhile, the widely used `{{#next_post}}{{/next_post}}` & `{{#prev_post}}{{/prev_post}}` helpers will work exactly as they did before, with no changes. The underlying implementation will be more efficient (hopefully improving page load times and reducing load on the database).

Finally, we're adding an extra parameter to the`{{#next_post}}{{/next_post}}` & `{{#prev_post}}{{/prev_post}}` to support fetching the next/prev post inside of a primary tag.

```
{{#next_post in="primary_tag"}}
      {{> "post-card"}}
{{/next_post}}

{{#prev_post in="primary_tag"}}
       {{> "post-card"}}
{{/prev_post}}
```

Things that need more testing:

- [x] does it work properly on sqlite3 vs mysql
- [x] is the order correct?
- [x] does it still properly filter out pages, and drafts, etc

### CODE REVIEW

File: core/server/helpers/prev_next.js [hunk 1]
```diff
@@ -13,20 +15,41 @@ var proxy = require('./proxy'),
...
+        };
+
+    if (_.get(options, 'hash.in') && options.hash.in === 'primary_tag' && _.get(post, 'primary_tag.slug')) {
+        apiOptions.filter += '+primary_tag:' + post.primary_tag.slug;
+    }
+
+    return apiOptions;
+};
+
+fetch = function fetch(options, data) {
```
Comment (kirrg001) [inline core/server/helpers/prev_next.js]: fetch(options, data)
buildApiOptions(post, options)

^ inconsistent argument order

File: core/server/helpers/prev_next.js [hunk 2]
```diff
@@ -44,21 +67,20 @@ fetch = function fetch(apiOptions, options, data) {
...
         logging.warn(data.error);
         return Promise.resolve();
     }
 
-    if (isPost(this) && this.status === 'published') {
-        apiOptions.slug = this.slug;
-        return fetch(apiOptions, options, data);
-    } else {
+    // Guard against trying to execute prev/next on previews, pages, or other resources
+    if (!isPost(this) || this.status !== 'published') {
```
Comment (kirrg001) [inline core/server/helpers/prev_next.js]: I think a page check is missing here. 
This was previously handled [here](https://github.com/TryGhost/Ghost/blob/master/core/server/models/post.js#L707).
Or we have to add this to our apiOptions above.

Comment (ErisDS) [reply core/server/helpers/prev_next.js]: @kirrg001 I can certainly add an extra check, but this is already handled by the same logic that makes sure we get posts by default when we do a browse call to the post API

The SQL query that gets used is:

```
{ method: 'select',
  options: {},
  timeout: false,
  cancelOnTimeout: false,
  bindings: 
   [ 'published',
     false,
     'yui2-javascript-carousel-an-update-about-version-2-8',
     '2010-10-10 17:56:20',
     1 ],
  __knexQueryUid: 'fa63caec-1b84-4bd3-a37f-7f36c4a11dc0',
  sql: 'select "posts".* from "posts" where ("posts"."status" = ? and "posts"."page" = ?) and ("posts"."slug" != ? and "posts"."published_at" <= ?) order by "posts"."published_at" DESC limit ?' }
```
OR, for primary tags:

```
{ method: 'select',
  options: {},
  timeout: false,
  cancelOnTimeout: false,
  bindings: 
   [ 'published',
     false,
     'starting-out-with-zend-framework',
     '2011-10-16 09:54:50',
     'code',
     0,
     'public',
     1 ],
  __knexQueryUid: '143723bf-fe4d-4e3c-85d8-f4490787b16c',
  sql: 'select "posts".* from "posts" left outer join "posts_tags" on "posts_tags"."post_id" = "posts"."id" left outer join "tags" on "posts_tags"."tag_id" = "tags"."id" where ("posts"."status" = ? and "posts"."page" = ?) and ("posts"."slug" != ? and "posts"."published_at" > ? and ("tags"."slug" = ? and "posts_tags"."sort_order" = ? and "tags"."visibility" = ?)) group by "posts"."id" order by "posts"."published_at" ASC limit ?' }
```

Comment (kirrg001) [reply core/server/helpers/prev_next.js]: But it has a different meaning or?
If you use `next_post` on a static page OR if you query for the next post which is NOT a page.

File: core/test/functional/routes/api/public_api_spec.js [hunk 3]
```diff
@@ -503,4 +504,70 @@ describe('Public API', function () {
...
     });
+
+    it('fetch the most recent post, then the prev, then the next should match the first', function (done) {
+        function createFilter(publishedAt, op) {
+            // This line deliberately uses double quotes because GQL cannot handle either double quotes
+            // or escaped singles, see TryGhost/GQL#34
+            return encodeURIComponent("published_at:" + op + "'" + publishedAt + "'");  // jscs:ignore
+        }
+
+        request.get(testUtils.API.getApiQuery(
```
Comment (kirrg001) [inline core/test/functional/routes/api/public_api_spec.js]: Did you know that [supertest](https://github.com/visionmedia/supertest#example) offers using `.then` chaining?

Comment (ErisDS) [reply core/test/functional/routes/api/public_api_spec.js]: I did not

Comment (kirrg001) [reply core/test/functional/routes/api/public_api_spec.js]: I discovered it last week while working on a service. I was 🎊.

**Review by kirrg001 — APPROVED** [review]

**Pull request merged**